microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigSourceProviderDefinition.java

```diff
@@ -65,30 +66,10 @@
     static AttributeDefinition[] ATTRIBUTES = { CLASS };
 
     protected ConfigSourceProviderDefinition(Registry<ConfigSourceProvider> providers) {
-        super(MicroProfileConfigExtension.CONFIG_SOURCE_PROVIDER_PATH,
-                MicroProfileConfigExtension.getResourceDescriptionResolver(MicroProfileConfigExtension.CONFIG_SOURCE_PROVIDER_PATH.getKey()),
-                new AbstractAddStepHandler(ATTRIBUTES) {
-                    @Override
-                    protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
-                        super.performRuntime(context, operation, model);
-                        String name = context.getCurrentAddressValue();
-                        ModelNode classModel = CLASS.resolveModelAttribute(context, model);
-                        if (classModel.isDefined()) {
-                            Class configSourceProviderClass = unwrapClass(classModel);
-                            try {
-                                providers.register(name, ConfigSourceProvider.class.cast(configSourceProviderClass.newInstance()));
-                            } catch (Exception e) {
-                                throw new OperationFailedException(e);
-                            }
-                        }
-                    }
-                }, new AbstractRemoveStepHandler() {
-                    @Override
-                    protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
-                        String name = context.getCurrentAddressValue();
-                        providers.unregister(name);
-                    }
-                });
+        super(new SimpleResourceDefinition.Parameters(MicroProfileConfigExtension.CONFIG_SOURCE_PROVIDER_PATH,
+                MicroProfileConfigExtension.getResourceDescriptionResolver(MicroProfileConfigExtension.CONFIG_SOURCE_PROVIDER_PATH.getKey()))
+                .setAddHandler(new ConfigSourceProviderDefinitionAddHandler(providers))
+                .setRemoveHandler(ReloadRequiredRemoveStepHandler.INSTANCE));
```

Comment:
using a `ReloadRequiredRemoveStepHandler.INSTANCE` will not be enough as you would also need to remove the current instance from the providers:

```java
new AbstractRemoveStepHandler() {		
                     @Override		
                     protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {		
                         String name = context.getCurrentAddressValue();		
                         providers.unregister(name);		
                     }		
                 });
```

Reply:
It sounds like we are missing an MSC service to do the registration/unregistration then, so the service stop does a remove during a reload.

Reply:
Brian is right, we need to have 3 more MSC service to take care of:
* ConfigSource loaded from class
* ConfigSource from Properties
* ConfigSourceProvide

All of them will be pretty similar to https://github.com/wildfly/wildfly/blob/master/microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/DirConfigSourceRegistrationService.java

Reply:
The reason all of that code was removed was because Brian said that if a reload is required, the handler shouldn't be changing the runtime state. Am I misunderstanding something here?

Reply:
I talked to Brian and have a better understanding of what's going on here. Change is in progress...